### DISCUSSION TITLE
Running Headless

### DISCUSSION THREAD

**Discussion by mikecasas** [oqtane/oqtane.framework#907] [Ideas] [answered]
Is there some setting to run Oqtane with just the backend?

**Comment by hishamco** [comment]
Nothing can stop you to use Oqtane as Source of Data, this could be done via built-in APIs

**Reply by mikecasas** [reply]
But the "client" piece would be deployed as well correct?

**Reply by hishamco** [reply]
Yes If I'm not wrong

/cc @sbwalker

**Reply by sbwalker** [reply]
For headless you should not need to deploy the client DLL as it only contains the Blazor components. You could test this using curl or postman and calling the API endpoints using valid URLs. However you would still need to implement some way to authenticate the user making the request as most of the APIs are protected in Oqtane.

**Reply by hishamco** [reply]
@sbwalker AFAIK `Runtime` is define in Clients.UI so if it's not deployed an exception will throw or we need to move the API in a proper place without introducing a breaking changes

**Reply by mikecasas** [reply]
Maybe handle it on the config.

**Reply by sbwalker** [reply]
@hishamco you are correct - the Runtime enum class in the Client project should be in the Shared project - as it is used in both the client and server.

**Reply by hishamco** [reply]
It's already in `appsettings.json` but we need it to be a strong type

**Reply by hishamco** [reply]
> @hishamco you are correct - the Runtime enum class in the Client project should be in the Shared project - as it is used in both the client and server.

I will push a PR keeping in mind backward compatibility and no breaking changes

**Comment by sbwalker** [comment, marked as answer]
@mikecasas it would be helpful to know what are you trying to accomplish...

**Reply by hishamco** [reply]
Seems he want to use the Oqtane APIs without Blazor UI

**Reply by mikecasas** [reply]
Long term, I want to run just the backend/apis in a container, and then serve the front end (the blazor web assembly) from some sort of web/file server. And obviously point the front end Blazor/Oqtane to the Backend container.

I know my use case may be unique.

**Reply by sbwalker** [reply]
I am not sure if you are understanding the client/server architecture for WebAssembly. The Oqtane Client is not something you would run in a container on its own. When running on Blazor WebAssembly a browser accesses the server and if the server is configured for WebAssembly it sends the Oqtane Client assembly to the browser which then loads it and downloads any additional assemblies which are needed. So if you intend to use the Oqtane UI then you need to include the Oqtane Client assembly on the server so that the browser can download it.

**Reply by mikecasas** [reply]
Yes, I should have said that I would like a front end container running nginx as my webserver serving up the oqtane client. Users would hit the front end and nginx would serve down the needed bits.

**Reply by sbwalker** [reply]
@mikecasas that will not work. As I mentioned above you would never run the Oqtane Client in its own container or on its own web server. It is not a "client" in the sense of being a standalone application. It is a "client" library which needs a host to run within. That host on WebAssembly is the web browser - not a web server.

**Reply by mikecasas** [reply]
I edited my comment above. Looking at it now, the wording was confusing.

I want to place the oqtane client on a web/file server for distribution purposes.

**Reply by PizzaConsole** [reply]
Not possible, due to the way some resources are dynamically loaded using the _Host.cshtml file. Whether you rune Oqtane as Webassembly or as server version it has to all run together on a server. You cannot decouple the client side from the server side. As .cshtml files are server rendered.

**Reply by sbwalker** [reply]
I provided the technical details for this on another thread "Early versions of Oqtane were configured to have an index.html in the Client project and an _host.cshtml in the Server project. In addition to these changes, you need to duplicate all Client static resources that are currently in the Server wwwroot to a wwwroot folder in the Client project. In order to support both Blazor hosting models with a single application, and avoid duplication of resources in both the Client and Server, we removed the wwwroot folder in the Client project and opted to bootstrapped WebAssembly using _host.cshtml. The other item to keep in mind is that some of the dynamic global JavaScript and CSS registration features are dependent on _host.cshtml as they rely on dynamic injection of elements to the initial page - which would be challenging to accomplish in a static index,html page. I am sure it would be possible to modify Oqtane to support a decoupled static site use case - with only minor degradation in features."

**Reply by sbwalker** [reply]
I should also mention that "running headless" usually means the ability to run a back-end services API without any specific client UI - this is why it is called "headless" because a client app is considered to be the "head". The initial discussion in this thread were focused on how you would configure Oqtane to run headless ( ie. just the server components ). In this model, in theory you could create a client front-end in Angular or React and still use the Oqtane back-end. The discussion then evolved to being how to deploy the client and server aspects of Oqtane in a physically separated model - which is not really related to "headless".

**Reply by mikecasas** [reply]
Just for "completeness" and to clarify what I was asking. I pushed the Oqtane Client to Github pages. It obviously didn't work because I did not have a backend, but at least the client "came up" and I was able to see the error messages.

In response to the other threads, I personally would trade off some of the "dynamic-ism", for better decoupling.

Way out there - I could see the the Oqtane.Server as a micro services or even a server less backend, where you can scale specific parts. The DAPR project (https://dapr.io/) is interesting with it's side car model. This could align with Oqtane's philosophy of not taking dependencies, but achieve nice features like pub-sub. 

I know I said alot, but just my 2 cents.

**Reply by PizzaConsole** [reply]
This was the same topic I posted here: #697 I will close mine as well as given the current state of Oqtane it will not be possible and would **possibly** cause ~~breaking changes~~ now if we were to try and change it. For this I have begun work on my own framework to achieve this functionality.

**Reply by sbwalker** [reply]
It is not clear to me how it would cause breaking changes? I mentioned above that I am sure it would be possible to modify Oqtane to support a decoupled static site use case.

**Reply by PizzaConsole** [reply]
It was just my initial thought given how the resources are dynamically added into the _host.cshtml file. My assessment carries no weight, as I have not spent much time investigating it.

**Comment by mikecasas** [comment]
I'll close this, as it's gotten off topic.

**Reply by thabaum** [reply]
Good question, good topic.  Thank you for the read.
